### PR TITLE
chore(metrics): refactor metrics v2 so it uses classes

### PR DESCRIPTION
**Pull request by Molter73**
**What type of PR is this?**

This is an alternative to the original implementation. Instead of using capturing lambdas that get called back after some additional work is done, we use classes and gather data during instantiation of the object. This approach should be a lot more straight forward and it also doesn't create unneeded objects when there's no need to do so. Another benefit of this approach is that metrics are bundled together in a very specific way, rather than depending on the ordering of definitions in an enum, which is a C idiom.

/kind cleanup

**Any specific area of the project related to this PR?**

/area libsinsp

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
At the time of creation, this refactoring is merely a proposal, any other changes are welcomed and the possibility of completely scrapping the PR in favor of a better alternative is also up there

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by Molter73** [comment]
/cc @incertum @FedeDP 
You both worked on this code, so I'd like your opinion on it 😄

File: userspace/libsinsp/metrics_collector.h
Comment (incertum) [inline userspace/libsinsp/metrics_collector.h:51]: Any chance to keep the comments somewhere? Perhaps when we emplace_back the metrics? Just so the next developer understands better where they come from. A handful of the metrics benefit from the comments for sure.

**Review by incertum — COMMENTED** [review]
Thanks Mauro!

Yes these changes work for me. We would need a Falco PR as well for the changes related to `libs::metrics::libsinsp_metrics::new_metric`. Would you mind opening that one as well?

@FedeDP WDYT? It also needs to be coordinated with your open PR.

**Comment by incertum** [comment]
Also CC @sgaist, thanks!

**Comment by FedeDP** [comment]
> @FedeDP WDYT? 

I really like it too! 
> It also needs to be coordinated with your open PR.

Yes; i can rebase if needed! :)

File: userspace/libsinsp/metrics_collector.h
Comment (FedeDP) [reply userspace/libsinsp/metrics_collector.h:51]: Agree with Melissa!

Comment (Molter73) [reply userspace/libsinsp/metrics_collector.h:51]: Sounds good, I'll find a new home for them!

**Comment by Molter73** [comment]
> Yes these changes work for me. We would need a Falco PR as well for the changes related to `libs::metrics::libsinsp_metrics::new_metric`. Would you mind opening that one as well?

I'll get that up and running as soon as I can, it might take a minute since I've never actually compiled Falco as a whole 😅

**Comment by Molter73** [comment]
> Yes; i can rebase if needed! :)

Also, since I might take sometime to put together the falco counter part to this PR, feel free to merge yours and I'll rebase this one, think perf testing is more important too.

**Comment by FedeDP** [comment]
Let's see what get merged first ahah

File: userspace/libsinsp/metrics_collector.h
Comment (Molter73) [reply userspace/libsinsp/metrics_collector.h:51]: I added the comments back, but instead of doing so where we emplace the variables, I added them next to the variables that store the values. I feel this closer resembles the original positions for the comments and has the benefit that it adds doxygen documentation for those fields.
I also used [member groups](https://www.doxygen.nl/manual/grouping.html#memgroup) to prevent duplicating some comments, this one is a bit nastier and I'm not fully convinced it's the right way to go about it, let me know what you think.

**Comment by Molter73** [comment]
I'm still working on the main repo counterpart for this PR, think I have a working change but have to test it still, should have it up for review soonish.

**Comment by incertum** [comment]
We merged Fede's PR first. @Molter73 when you get a chance could you rebase and once the Falco PR is up we can merge it quickly.

**Review by FedeDP — APPROVED** [review]
Like it!
/approve

**Comment by Molter73** [comment]
/hold

Still need to open the main repo counterpart.

**Comment by Molter73** [comment]
I've opened falcosecurity/falco#3265 as the counter part to this PR, if everything looks good over there, we can go ahead and merge this.

**Pull request merged**
